### ISSUE TITLE
Login endpoint blocked by geetest captcha

### ISSUE THREAD

**Issue by raidensakura** [raidensakura/hoyolab-auto-login#6]
Currently the script is impacted by Hoyoverse integrating Geetest into their login endpoint, the request will be replied with OK response but with extra geetest verification payload.

I've already added error handling for when the script is blocked by the captcha but the login script will not work until either a solution is found to bypass it, or Hoyoverse remove their captcha system altogether (which is very unlikely).

As of now from my testing, Only Genshin's login endpoint seem to be affected, but this may only be a matter of time before it gets rolled out to other game's endpoint as well.

```md
[2023-06-18 22:37:19] [INFO] Cookie 1/2 OK, Hello R****ia
[2023-06-18 22:37:23] [INFO] Checking in Raiden ♫ for Honkai Impact 3 (UID: xxxx7789)
[2023-06-18 22:37:25] [INFO] Checking in 玩家202423125 for Tears of Themis (UID: xxxx23125)
[2023-06-18 22:37:25] [INFO] Checking in Raiden for Genshin Impact (UID: xxxx49533)
[2023-06-18 22:37:26] [ERROR] Blocked by geetest captcha :(
{"risk_code": 5001, "gt": "a280f1e5745a37ae91c65fc745fba71c", "challenge": "d1cc36a084d1ae39531eaeb019876bbc", "success": 1, "is_risk": true}
[2023-06-18 22:37:26] [INFO] Checking in Raiden for Honkai: Star Rail (UID: xxxx64080)
[2023-06-18 22:37:27] [INFO] Successfully sent Discord embed
[2023-06-18 22:37:27] [INFO] Cookie 2/2 OK, Hello T****s_
[2023-06-18 22:37:31] [INFO] Checking in Tigris for Genshin Impact (UID: xxxx69263)
[2023-06-18 22:37:32] [ERROR] Blocked by geetest captcha :(
{"risk_code": 5001, "gt": "a280f1e5745a37ae91c65fc745fba71c", "challenge": "ad4a4eaccf9ff70b1d2cade6fe40963b", "success": 1, "is_risk": true}
[2023-06-18 22:37:32] [INFO] Sleeping for a day..
```

**Comment by BuIlDaLiBlE** [comment]
This is very sad news. Would be nice to have a feature to ping the user in Discord when the captcha prevents claiming, perhaps by mapping UID to Discord ID and ping the user by that ID in the embed?

**Comment by raidensakura** [comment]
> This is very sad news. Would be nice to have a feature to ping the user in Discord when the captcha prevents claiming, perhaps by mapping UID to Discord ID and ping the user by that ID in the embed?
Added this in [5f7eebc](https://github.com/raidensakura/hoyolab-auto-login/commit/5f7eebc3bba75583816ee0d0018d6ecb30d4d9c8), you can now append `DISCORD_ID=1234567890; rest_of_cookie=; ...` in your cookie and the webhook will notify by pinging the ID in case of a failure.

**Comment by raidensakura** [comment]
6a2643e Added support for solving Geetest using [2captcha](https://2captcha.com/). To do this, simply add `2CAPTCHA_API=your_api_key` to your env variables.

This require registration and payment. It costs around $3 for 1000 captcha solving requests, which each game's login will consume 1. For metrics, assuming you play 3 games and they all have geetest enabled, you will be good for about 300 days before needing to pay again.

I'll keep this issue open since I don't really consider it solved since it's just technically offloading the captcha solving to an external service. But it should provide a solution at least for those who are willing to fork some cash.

**Comment by raidensakura** [comment]
> P.S. Is there a reason why we added the discord token to the cookie env variable instead of just having a `DISCORD_ID` env variable on its own?

Needed a way to bind the ID to a cookie, the embed message get sent once per cookie and I couldn't think of a clean way to handle in the case of multiple cookie setup with only a certain cookie wanting to be pinged
ie: 
```
cookie=cookie1#cookie2#cookie3#cookie4#cookie5
discord_id=user1,user3,user5
                 ^^^^  ^^^^
```
Always open for PR if there's a solution better for the user

**Comment by Equ1no0x** [comment]
> Added support for solving Geetest using [2captcha](https://2captcha.com/). To do this, simply add `2CAPTCHA_API=your_api_key` to your env variables.

I and a friend got it set up to use 1 key, we each pay around $6 a year for all accounts (both using the same API token) a pretty good deal.
I don't mind it, but if a free solution does appear, it would be awesome.

**Comment by raidensakura** [comment]
Closing this as completed since I've read multiple confirmation from other users that they're no longer receiving the captcha, including from my own use.
